prow/cmd/phaino/local.go

```diff
@@ -0,0 +1,339 @@
+		entrypoint = args[0]
+		args = args[1:]
+	}
+	if entrypoint == "" && decoration != nil {
+		return nil, errors.New("decorated jobs must specify command and/or args")
+	}
+	if entrypoint != "" {
+		localArgs = append(localArgs, "--entrypoint="+entrypoint)
+	}
+
+	for _, env := range container.Env {
+		localArgs = append(localArgs, "-e", env.Name+"="+env.Value)
+	}
+
+	priv, err := checkPrivilege(container, allowPrivilege)
+	if err != nil {
+		return nil, err
+	}
+	if priv {
+		localArgs = append(localArgs, "--privileged")
+	}
+
+	if container.Resources.Requests != nil {
+		// TODO(fejta): https://docs.docker.com/engine/reference/run/#runtime-constraints-on-resources
+		log.Warn("Ignoring resource requirements")
+	}
+
+	for _, mount := range container.VolumeMounts {
+		vol := volume(*pj.Spec.PodSpec, mount.Name)
+		if vol == nil {
+			return nil, fmt.Errorf("mount %q missing associated volume", mount.Name)
+		}
+		if vol.EmptyDir != nil {
+			localArgs = append(localArgs, "-v", mount.MountPath)
+		} else {
+			local, err := readMount(mount)
+			if err != nil {
+				return nil, fmt.Errorf("bad mount %q: %v", mount.Name, err)
+			}
+			arg := local + ":" + mount.MountPath
+			if mount.ReadOnly {
+				arg += ":ro"
+			}
+			localArgs = append(localArgs, "-v", arg)
+		}
+	}
+
+	var workingDir string
+
+	if decoration != nil {
+		var refs []prowapi.Refs
+		if pj.Spec.Refs != nil {
+			refs = append(refs, *pj.Spec.Refs)
+		}
+		refs = append(refs, pj.Spec.ExtraRefs...)
+		for _, ref := range refs {
+			path := pathAlias(ref)
+			repo, err := readRepo(path)
+			if err != nil {
+				return nil, fmt.Errorf("bad %q repo: %v", path, err)
+			}
+			dest := filepath.Join("/go/src", path)
+			if workingDir == "" {
+				workingDir = dest
+			}
+			localArgs = append(localArgs, "-v", repo+":"+dest)
+
+		}
+	}
+	if workingDir == "" {
+		workingDir = "/random-location"
+		localArgs = append(localArgs, "-v", workingDir)
+	}
+	localArgs = append(localArgs, "-w", workingDir)
+
+	image := pj.Spec.PodSpec.Containers[0].Image
+	localArgs = append(localArgs, image)
+	localArgs = append(localArgs, args...)
+	return localArgs, nil
+}
+
+func printArgs(localArgs []string) {
+	base := len(baseArgs)
+	for i, a := range localArgs {
+		if i < base {
+			fmt.Printf("%q ", a)
+		} else {
+			fmt.Printf("\\\n %q ", a)
+		}
+	}
+	fmt.Println()
+}
+
+func start(args []string) (*exec.Cmd, error) {
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+
+	return cmd, cmd.Start()
+}
+
+func abort(ctx context.Context, log *logrus.Entry, cmd *exec.Cmd) error {
+	if err := cmd.Process.Signal(os.Interrupt); err != nil {
+		log.WithError(err).Warn("Interrupt error")
+	} else {
+		log.Warn("Interrupted")
+	}
+	ch := make(chan error)
+	go func() {
+		defer close(ch)
+		ch <- cmd.Wait()
+	}()
+	select {
+	case err := <-ch:
+		return err
+	case <-ctx.Done():
+		if err := cmd.Process.Kill(); err != nil {
```

Comment:
Is this actually going to stop the job running, or just detach the client from it?

Reply:
Added a TODO. Assuming killing a `docker run --rm=true` process doesn't also kill the container (quite likely) it will be mildly annoying to get the container id first.